checks/remotesettings/certificates_expiration.py

```diff
@@ -0,0 +1,86 @@
+"""
+Signature certificates should expire in a minimum number of days.
+
+The list of collections whose certificate expires too soon is returned, along
+with the expiration date and x5u URL.
+"""
+import asyncio
+import logging
+from datetime import datetime
+
+import cryptography
+import cryptography.x509
+import requests
+from cryptography.hazmat.backends import default_backend as crypto_default_backend
+from poucave.typings import CheckResult
+
+from .utils import KintoClient as Client
+
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_collection_metadata(server_url, entry):
+    client = Client(
+        server_url=server_url, bucket=entry["bucket"], collection=entry["collection"]
+    )
+    return client.get_collection(_expected=entry["last_modified"])["data"]
+
+
+def fetch_certificate_expiration(x5u: str) -> datetime:
+    resp = requests.get(x5u)
+    cert_pem = resp.text.encode("utf-8")
+    cert = cryptography.x509.load_pem_x509_certificate(
+        cert_pem, crypto_default_backend()
+    )
+    return cert.not_valid_after
+
+
+async def run(server: str, min_remaining_days: int) -> CheckResult:
+    loop = asyncio.get_event_loop()
+
+    client = Client(server_url=server, bucket="monitor", collection="changes")
+    entries = client.get_records()
+
+    futures = [
+        loop.run_in_executor(None, fetch_collection_metadata, server, entry)
+        for entry in entries
+    ]
+    results = await asyncio.gather(*futures)
+    entries_metadata = zip(entries, results)
+
+    x5us = list(
+        {
+            metadata["signature"]["x5u"]
+            for metadata in results
+            if "signature" in metadata
+        }
+    )
+    futures = [
+        loop.run_in_executor(None, fetch_certificate_expiration, x5u) for x5u in x5us
+    ]
+    results = await asyncio.gather(*futures)
```

Comment:
The fact that we have more than one `asyncio.gather` seems a little bit like a code smell to me. At first I thought this was an optimization because some certificates are repeated, but it seems like we don't do that either. (If we do, then we should put a comment, something like `Gather x5us first to deduplicate`.)

It seems to me that we don't need the intermediate result. Could this maybe be cleaner if we chain the futures in some other way, like a function that returns the URL and its expiration as a pair?

Reply:
Yes we do deduplicate (the set syntax is not super visible), I added a comment.

Reply:
OK, thanks, I don't know why I didn't see it the first time through.